### PR TITLE
GPS CNAV message decoding support

### PR DESCRIPTION
**Pull request by valeri-atamaniouk**
Added capability to decode GPS CNAV message and extract the following information: CNAV message header, symbol delay.

### CODE REVIEW

**Comment by rge-exafore** [comment]
@swift-nav/firmware, could you please review and merge?

File: src/cnav_msg.c [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+ * CNAV decoder contains two Viterbi decoders that are used to estimate bit and
+ * message boundary.
+ *
+ * \param[out] dec Decoder structure.
+ *
+ * \return None
+ */
+void cnav_msg_decoder_init(cnav_msg_decoder_t *dec)
+{
+  static const signed char polynomial[2] = {
```
Comment (jacobmcnamee) [inline src/cnav_msg.c]: polynomial does not need to be statically allocated

File: src/cnav_msg.c [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+   * - M - Number of bits in the tail to ignore.
+   */
+  unsigned char tmp_bits[ (GPS_L2C_V27_DECODE_BITS + GPS_L2C_V27_DELAY_BITS +
+                           CHAR_BIT - 1) / CHAR_BIT];
+
+  v27_chainback_likely(&part->dec, tmp_bits,
+                       GPS_L2C_V27_DECODE_BITS + GPS_L2C_V27_DELAY_BITS);
+
+  /* Read decoded bits and add them to the decoded buffer */
+  u32 to_add = getbitu(tmp_bits, 0, GPS_L2C_V27_DECODE_BITS);
```
Comment (jacobmcnamee) [inline src/cnav_msg.c]: This breaks if GPS_L2C_V27_DECODE_BITS > 32. Can we add a comment or an assert or something?

File: src/cnav_msg.c (hunk 1 above)
Comment (valeri-atamaniouk) [inline src/cnav_msg.c]: I disagree. The coefficients never change and they have to be passed as an array. There is no reason to allocate they array in heap and programmatically fill it.

Comment (jacobmcnamee) [inline src/cnav_msg.c]: I mean it is not required - the Viterbi decoder will work fine either way.

If we're concerned with speed or memory, it might be good to preinitialize and share a single v27_poly_t between all the CNAV decoders.

**Comment by jacobmcnamee** [comment]
A couple // comments in the header file and test file, but otherwise lgtm.

File: include/libswiftnav/cnav_msg.h [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  u32  tow;    //!< GPS ToW in 6-second units. Multiply to 6 to get seconds.
+  bool alert;  //!< CNAV message alert flag
+} cnav_msg_t;
+
+/**
+ * GPS CNAV decoder component.
+ * This component controls symbol decoding string.
+ *
+ * @sa cnav_msg_decoder_t
+ */
```
Comment (fnoble) [inline include/libswiftnav/cnav_msg.h]: Comments and documentation are great! Just a small note that we use a slightly different style for our Doxygen comments. Not a big deal but good to be consistent.

https://swiftnav.hackpad.com/Engineering-Playbook-fKcJhOl85sD#:h=Doxygen

File: src/cnav_msg.c [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+/** \defgroup GPS_L2 GPS L2
+ * GPS L2 operations
+ * \{ */
+/** \defgroup gps_cnav_decoder Decoder
+ * GPS L2C CNAV message decoding.
+ * \{ */
+/*
+ * Block Viterbi decoding parameters.
+ */
+/** Viterbi decoder inverted polynome A */
```
Comment (fnoble) [inline src/cnav_msg.c]: polynome -> polynomial

File: src/cnav_msg.c [hunk 5]
```diff
@@ -0,0 +1,515 @@
...
+/** GPS L2C CNAV message length in bits */
+#define GPS_CNAV_MSG_LENGTH         (300)
+/** GPS LC2 CNAV CRC length in bits */
+#define GPS_CNAV_MSG_CRC_LENGTH     (24)
+/** GPS L2C CNAV message payload length in bits */
+#define GPS_CNAV_MSG_DATA_LENGTH    (GPS_CNAV_MSG_LENGTH-GPS_CNAV_MSG_CRC_LENGTH)
+/** GPS L2C CNAV message lock detector threshold */
+#define GPS_CNAV_LOCK_MAX_CRC_FAILS (10)
+
+// forward declarations for test access
```
Comment (fnoble) [inline src/cnav_msg.c]: `//` -> `/* */`

File: tests/check_cnav.c [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+#include <stdlib.h>
+#include <math.h>
+#include <check.h>
+
+#include "check_utils.h"
+
+#define G1 0x79 // 0171
+#define G2 0x5B // 0133
+
+// internal CNAV declarations for test access
```
Comment (fnoble) [inline tests/check_cnav.c]: I worry this could cause unexpected behavior if you forget to keep the prototypes in sync. Another option is to declare the functions static in `cnav_msg.c` and then in this test file do a `#include <cnav_msg.c>` (note .c not .h !)

Anyway, I don't mind too much which solution you choose. This stack overflow post has more ideas:
http://stackoverflow.com/questions/593414/how-to-test-a-static-function

File: src/edc.c [hunk 7]
```diff
@@ -77,6 +77,46 @@ u32 crc24q(const u8 *buf, u32 len, u32 crc)
   return crc;
 }
 
+/**
+ * Computes CRC-24Q for left-aligned bit message.
+ * This function is used for left-aligned bit messages, for example SBAS and
+ * GPS CNAV.
+ * GPS message is 300 bits total, but 276 bits without CRC. It takes 34.5
+ * 8-bit bytes, and when computing CRC the message has to be padded with zero
+ * bits.
```
Comment (fnoble) [inline src/edc.c]: @jacobmcnamee : how did you deal with this for SBAS? Did you use the existing CRC function?

File: src/cnav_msg.c [hunk 8]
```diff
@@ -0,0 +1,515 @@
...
+ *
+ * \param[in] buf   Pointer to buffer head.
+ * \param[in] bits  Number of bytes in the buffer.
+ * \param[in] shift Number of bits for left shift operation.
+ *
+ * \return None
+ *
+ * \private
+ */
+void _cnav_bitlshift(void *buf, size_t size, u32 shift)
```
Comment (fnoble) [inline src/cnav_msg.c]: This function seems generally useful outside of CNAV, maybe move to `bits.c`?

File: src/cnav_msg.c [hunk 9]
```diff
@@ -0,0 +1,515 @@
...
+ * \param[in]     src       Source buffer.
+ * \param[in]     src_index Source bit index.
+ * \param[in]     count     Number of bits to copy.
+ *
+ * \return None
+ *
+ * \private
+ */
+void _cnav_bitcopy(void *dst, u32 dst_index,
+                   const void *src, u32 src_index, u32 count)
```
Comment (fnoble) [inline src/cnav_msg.c]: Also seems generally useful, move to `bits.c`?

File: src/cnav_msg.c [hunk 10]
```diff
@@ -0,0 +1,515 @@
...
+/**
+ * Invert message bits in the buffer.
+ *
+ * The method inverts bits of the decoded data.
+ *
+ * \param[in,out] part Decoder component with a message buffer.
+ *
+ * \return None
+ */
+void _cnav_msg_invert(cnav_v27_part_t *part)
```
Comment (fnoble) [inline src/cnav_msg.c]: Also a candidate to be turned into a generic invert buffer function in `bits.c`?

File: src/edc.c (hunk 7 above)
Comment (jacobmcnamee) [inline src/edc.c]: We didn't really get that far. Vlad's SBAS support changes for piksi_firmware or libswiftnav might have something if you are really curious.

File: src/cnav_msg.c [hunk 11]
```diff
@@ -0,0 +1,515 @@
...
+  dec->part2.init = true;
+  _cnav_add_symbol(&dec->part2, 0x80);
+}
+
+/**
+ * Adds a received symbol to decoder.
+ *
+ * \param[in,out] dec    Decoder object.
+ * \param[in]     symbol Symbol value probability, where 0x00 - 100% of 0,
+ *                       0xFF - 100% of 1.
```
Comment (fnoble) [inline src/cnav_msg.c]: Might want to double check, I thought the `symbol` was passed as a signed number, i.e. 2's complement

File: tests/check_cnav.c [hunk 12]
```diff
@@ -0,0 +1,389 @@
...
+
+#include <time.h>
+#include <stdlib.h>
+#include <math.h>
+#include <check.h>
+
+#include "check_utils.h"
+
+#define G1 0x79 // 0171
+#define G2 0x5B // 0133
```
Comment (fnoble) [inline tests/check_cnav.c]: `//` -> `/* */` in a few places

**Comment by fnoble** [comment]
@valeri-atamaniouk : looking really good! I made a few comments but mostly minor.

I don't know if you have seen this but we have code coverage testing available here:
https://coveralls.io/builds/4852544/source?filename=src%2Fcnav_msg.c

Looks like there are a couple edge conditions currently missed that it would be good to hit.

Just address those comments and test coverage and we'll be good to merge.

File: include/libswiftnav/cnav_msg.h (hunk 3 above)
Comment (valeri-atamaniouk) [inline include/libswiftnav/cnav_msg.h]: Docs updated. No more C++ - style comments / documentation blocks

**Comment by valeri-atamaniouk** [comment]
Please do not merge. Some newer tests don't pass.

**Comment by valeri-atamaniouk** [comment]
@fnoble All comments addressed: code refactored, new unit-tests, coverage ok.
